### PR TITLE
test: group executed tests within the same directory

### PR DESCRIPTION
**Pull request by furszy**
Solving https://github.com/bitcoin/bitcoin/pull/31000#discussion_r1836348991.

Modifies the default datadir path for unit/bench/fuzz tests, grouping
all tests within a single directory labeled by the current timestamp.

Replicating the functional test framework behavior.

Directory structure update:
Before: ```tmp/test_common bitcoin/test_name/current_time```
After: ```tmp/test_common bitcoin/current_time/test_name```

### CODE REVIEW

**Review by pablomartin4btc — COMMENTED** [review]
ACK ce0645c312d0395627f1497140108cfe91473bc2

> Groups all tests executed within each binary call under a single directory prefixed by the current time. Replicating the function test framework behavior.

It makes sense to me. I'll play a bit with it just to see how it looks.

**Review by hodlinator — COMMENTED** [review]
Concept ACK ce0645c312d0395627f1497140108cfe91473bc2

Only case I can think of where having *time* segment before `test_name` in the path would be annoying is if one was re-running 2 specific tests a variable amount of times. This PR makes it so that one has to "open every door" (*time* segment) to see if files for test-A was there or for test-B. In general I think this way is a net-positive though.

Good to see the `-testdatadir` description updated, sorry for not catching it in #31000.

nit: Would prefer `current_time` replaced with `start_time` (or `process_start_time`) in the PR summary.

File: src/test/util/setup_common.cpp [hunk 1]
```diff
@@ -138,8 +141,7 @@ BasicTestingSetup::BasicTestingSetup(const ChainType chainType, TestOpts opts)
 
     const std::string test_name{G_TEST_GET_FULL_NAME ? G_TEST_GET_FULL_NAME() : ""};
     if (!m_node.args->IsArgSet("-testdatadir")) {
-        const auto now{TicksSinceEpoch<std::chrono::nanoseconds>(SystemClock::now())};
-        m_path_root = fs::temp_directory_path() / TEST_DIR_PATH_ELEMENT / test_name / util::ToString(now);
+        m_path_root = fs::temp_directory_path() / TEST_DIR_PATH_ELEMENT / util::ToString(g_start_time) / test_name;
```
Comment (hodlinator) [inline src/test/util/setup_common.cpp:144]: nit: We are recalculating `util::ToString()` for the same value every time we pass by this. Could do it once on line 80.
```suggestion
        m_path_root = fs::temp_directory_path() / TEST_DIR_PATH_ELEMENT / g_start_time / test_name;
```

**Comment by furszy** [comment]
> Only case I can think of where having time segment before test_name in the path would be annoying is if one was re-running 2 specific tests a variable amount of times.

For that we could introduce a `-repeat` arg, and suffix the dir with the loop round number.

**Comment by hodlinator** [comment]
> > Only case I can think of where having time segment before test_name in the path would be annoying is if one was re-running 2 specific tests a variable amount of times.
> 
> For that we could introduce a `-repeat` arg, and suffix the dir with the loop round number.

Maybe at that stage it's acceptable to require unique `-testdatadir` being passed in for every run.

File: src/test/util/setup_common.cpp (hunk 1 above)
Comment (furszy) [reply src/test/util/setup_common.cpp]: Sure. Done as suggested.

File: src/test/util/setup_common.cpp [hunk 2]
```diff
@@ -76,6 +76,9 @@ const std::function<std::string(const char*)> G_TRANSLATION_FUN = nullptr;
 
 constexpr inline auto TEST_DIR_PATH_ELEMENT{"test_common bitcoin"}; // Includes a space to catch possible path escape issues.
 
+/* Used to group tests executed within the same binary run under the same directory */
+static uint64_t g_start_time = TicksSinceEpoch<std::chrono::nanoseconds>(SystemClock::now());
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/test/util/setup_common.cpp]: Sure. Done as suggested. Thanks.

**Comment by furszy** [comment]
> > > Only case I can think of where having time segment before test_name in the path would be annoying is if one was re-running 2 specific tests a variable amount of times.
> > 
> > 
> > For that we could introduce a `-repeat` arg, and suffix the dir with the loop round number.
> 
> Maybe at that stage it's acceptable to require unique `-testdatadir` being passed in for every run.

For sure. That wouldn't be hard to do neither.

**Review by hodlinator — APPROVED** [review]
ACK b7f40e8094e00c13a6c847340645b16704fed63b

Ran `ctest -j 20 --test-dir build` and `build/src/test/test_bitcoin` and inspected activity with `lsof |grep "/tmp.*bitcoin`. The former will create a new process and hence *time* directory for every test, but run them in parallel. The latter will run all tests sequentially in-process, resulting in only one new *time* directory.

File: src/test/util/setup_common.cpp [hunk 3]
```diff
@@ -76,6 +76,9 @@ const std::function<std::string(const char*)> G_TRANSLATION_FUN = nullptr;
 
 constexpr inline auto TEST_DIR_PATH_ELEMENT{"test_common bitcoin"}; // Includes a space to catch possible path escape issues.
 
+/* Used to group tests executed within the same binary run under the same directory */
+static const std::string g_init_time{util::ToString(TicksSinceEpoch<std::chrono::nanoseconds>(SystemClock::now()))};
+
```
Comment (rkrux) [inline src/test/util/setup_common.cpp:81]: > Replicating the functional test framework behavior.

Should we use a prettier format here just like done in the functional tests? With this change the directory name is in plain nanoseconds like below.

```
drwx------    3 rkrux  staff    96B Nov 15 11:09 1731649167593349000
drwx------    3 rkrux  staff    96B Nov 15 11:11 1731649280506089000
```

Whereas for functional tests, the name is a little easier to read by using a date time format: https://github.com/bitcoin/bitcoin/blob/master/test/functional/test_runner.py#L476
```
drwxr-xr-x   10 rkrux  staff   320B Nov 15 12:25 test_runner_?_🏃_20241115_120930
drwxr-xr-x    7 rkrux  staff   224B Nov 15 15:12 test_runner_?_🏃_20241115_150614
```

**Review by rkrux — APPROVED** [review]
Concept ACK b7f40e8094e00c13a6c847340645b16704fed63b

Successful make and unit tests. Used `ctest --test-dir build -j 13` to run the unit tests and checked for the new directory setup in $TMPDIR, which unintuitively on OSX is not `/tmp`. :)

File: src/test/util/setup_common.cpp (hunk 3 above)
Comment (rkrux) [reply src/test/util/setup_common.cpp:81]: > Think of each unit test as a separate child class inheriting from the testing setup class and implementing a run() method. Each one constructs a new instance of the setup class, thereby creating a new rand_str.

One thing that I noticed is that there are several such (`1731649167593349000`) time directories when the unit tests are run, each having around one test_name directory inside. Whereas for the functional tests, a dir such as `test_runner_?_🏃_20241115_150614` encapsulates all of them.

**Comment by furszy** [comment]
Hmm, okay. Since the inclusion of CTest, we can run unit tests in parallel, but they are executed in standalone processes, which makes this approach impractical. Maybe we could address this at the CTest level, but I don’t have much time to spend on it.
If someone wants to tackle it, all yours. Up for grabs.

**Pull request closed without merging**